### PR TITLE
some ToDo s to make things better

### PR DESCRIPTION
**Pull request by DvonHolten**
- add a ToDo proposal avoid recompiling regex (twice!) on every use
- add a ToDo proposal to save a lot (?) of memory
- remove some trailing blanks

### CODE REVIEW

**Comment by THausherr** [comment]
Feel free to add a commit about a compiled constant regex. I don't have an opinion on the other topic.

**Comment by DvonHolten** [comment]
ok, regex coming.
the flag-thing should have some investigation on how many objects are created. is it 10 or 10000?

**Comment by xzel23** [comment]
Nice find.

For the flag thing, how about just changing isValid to this? 
```
    public boolean isValid()
    {
        return value != Long.MAX_VALUE && value != Long.MIN_VALUE;
    }
```

This would also make the constructor and where it is called a little bit shorter because there'd be no need to pass the `valid`parameter.

And the `STATIC` array could just be initialised in a static block, IMHO that wouldn't cost too much memory, and get(long) would be much simpler:

```
    public static COSInteger get(long val)
    {
        if (LOW <= val && val <= HIGH)
        {
            return STATIC[(int) val - LOW];
        }
        return new COSInteger(val, true);
    }
```

`getInvalid(boolean maxValue)` can also be removed and we use just
```
    protected static final COSInteger OUT_OF_RANGE_MAX = new COSInteger(Long.MAX_VALUE);
    protected static final COSInteger OUT_OF_RANGE_MIN = new COSInteger(Long.MIN_VALUE);
```

UPDATE: there would of course be a little semantic change because in the current implementation it's possible to `get(Long.MAX_VALUE)` to obtain a valid `COSInteger` with that value. But that looks rather like a bug in the current implementation to me because we'd have two instances where `equals()`returns `true` while one is valid and the other is not. Someone with more insight into the PDF spec has to decide if the current semantics though weird are mandated by the spec or not.

**Comment by xzel23** [comment]
@THausherr What do you think about the suggested change? I just "bought" the PDF spec yesterday (you have to buy it, but the price is 0 USD), and I couldn't find anything about COInteger to require Long.MIN_VALUE and Long.MAX_VALUE to map to valid COSInteger values.

@DvonHolten if Tilman and/or some other maintainer agrees to the proposed change, do you want to make the flag related change yourself or should I create a PR for that?

**Comment by THausherr** [comment]
I'm not deep enough into the COSInteger code, it is related to PDFBOX-5176 and any change (except removing `getInvalid()` because of https://twitter.com/unclebobmartin/status/1114137614377005057 ) should be checked with the file from there and with the file from the earlier issue. PDFBOX-5176 was created after running a regression test on 250000 PDF files.

**Comment by DvonHolten** [comment]
i can do it - check, where the invalid things are coming from and where they are checked. if all the Long-values must be available, a subclass may do it.
and i have a closer look at those 250000 pdf -files in PDFBOX-5176. i assume, they are just created to check the parser - not to render them?

**Comment by THausherr** [comment]
You don't have to bother with the 250000 files, this is our regression test corpus and we run these once on a dedicated server before doing a release. Just test the files from PDFBOX-5176 and the one from the related issue. I mentioned the 250000 files to show how unlikely this is.

File: fontbox/src/main/java/org/apache/fontbox/ttf/GlyphSubstitutionTable.java [hunk 1]
```diff
@@ -75,6 +76,13 @@ public class GlyphSubstitutionTable extends TTFTable
 
     private GsubData gsubData;
 
+    /**
+     * The regex represents 4 'word characters' [a-zA-Z_0-9],
+     * see {@link java.util.regex.ASCII#WORD}<br>
+     * Note: the ' '-character is not matched!
+     */
+    static final Pattern wordPattern = Pattern.compile( "\\w{4}" );
```
Comment (xzel23) [inline fontbox/src/main/java/org/apache/fontbox/ttf/GlyphSubstitutionTable.java]: The name should be all uppercase for constants (static final). It should also be private.

Comment (xzel23) [reply fontbox/src/main/java/org/apache/fontbox/ttf/GlyphSubstitutionTable.java]: What do you think about making this a predicate?

```
    private static final Predicate<String> IS_FOUR_LETTERS = Pattern.compile( "\\w{4}" ).asMatchPredicate();
```

Then the test would be `IS_FOUR_LETTERS.test(featureTags[i])`.

File: fontbox/src/main/java/org/apache/fontbox/ttf/GlyphSubstitutionTable.java [hunk 2]
```diff
@@ -799,10 +806,11 @@ public int getSubstitution(int gid, String[] scriptTags, List<String> enabledFea
     }
 
     /**
-     * For a substitute-gid (obtained from {@link #getSubstitution(int, String[], List)}), retrieve the original gid.
-     *
-     * Only gids previously substituted by this instance can be un-substituted. If you are trying to unsubstitute before
-     * you substitute, something is wrong.
+     * For a substitute-gid (obtained from {@link #getSubstitution(int, String[], List)}),
+     * retrieve the original gid.
+     * <br>
```
Comment (xzel23) [inline fontbox/src/main/java/org/apache/fontbox/ttf/GlyphSubstitutionTable.java]: Usually `<p>` is used in Javadoc to start a new paragraph.

**Review by xzel23 — COMMENTED** [review]
Note that I am not an official project member, so I just left some comments. I think the name of the matcher should be changed and made private, the comment about using a Predicate is just a suggestion that the maintainers may or may not like.

File: fontbox/src/main/java/org/apache/fontbox/ttf/GlyphSubstitutionTable.java (hunk 1 above)
Comment (THausherr) [reply fontbox/src/main/java/org/apache/fontbox/ttf/GlyphSubstitutionTable.java]: yeah, can be done for the trunk (needs jdk11). I'll use the previous suggestion for 2.0 and 3.0.

**Review by xzel23 — COMMENTED** [review]
Looks good to me now. Thank you!

**Comment by THausherr** [comment]
I've now committed the javadoc changes and the pattern change.

**Pull request closed without merging**